### PR TITLE
RUBY-3109 remove all references to dots/dollars

### PR DESCRIPTION
**Pull request by Neilshweky**
Some followup questions about this: 
- do we need to deprecate this in 4.15(16?)
- do we want to leave this functionality for the future?
- If not, can we remove the BSON::Config module?

Also in this PR, renamed DBRef argument error.

### CODE REVIEW

**Comment by p-mongo** [comment]
There are many conflicting files in this PR?

File: ext/bson/read.c [hunk 1]
```diff
@@ -143,7 +143,7 @@ VALUE rb_bson_byte_buffer_get_bytes(VALUE self, VALUE i)
 }
 
 VALUE pvt_get_boolean(byte_buffer_t *b){
-  VALUE result;
+  VALUE result = 0;
```
Comment (p-mongo) [inline ext/bson/read.c:146]: I don't think you can do this. If you want to silence the compiler warning, you can assign `Qnil` to `result` immediately before the `pvt_raise_decode_error` call in that branch only.

Comment (p-mongo) [reply ext/bson/read.c:146]: And add a comment that the assignment is to pacify the compiler.

**Comment by Neilshweky** [comment]
> There are many conflicting files in this PR?

ooh I just merged 1806, that must be why... I will rebase.

**Comment by Neilshweky** [comment]
@p-mongo  conflicts are gone!

**Comment by p-mongo** [comment]
This change should be announced for 5.0. We don't currently have release notes in bson-ruby, so for now the note can be added to https://jira.mongodb.org/browse/RUBY-1604 as a simple list of changes.

If you like you can add a deprecation to 4.x. My guess is the change will affect virtually no applications. If it's easy to add the deprecation, I'd say go for it, if it's a meaningful amount of work this can be skipped in my opinion.


> do we want to leave this functionality for the future?

My guess is it won't be used in any visible future. My vote is no.

> If not, can we remove the BSON::Config module?

Good question. It might be useful but having it be there and be empty is somewhat silly. @comandeo ?

File: ext/bson/read.c (hunk 1 above)
Comment (Neilshweky) [inline ext/bson/read.c:146]:
```suggestion
  VALUE result;
```

Comment (Neilshweky) [reply ext/bson/read.c]: I just got rid of it.

File: lib/bson/error/invalid_dbref_argument.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+module BSON
+  class Error
+    # Raised when trying to create a BSON::DBRef from an object that is an invalid DBRef.
+    #
+    # @api private
```
Comment (p-mongo) [inline lib/bson/error/invalid_dbref_argument.rb:7]: Exception classes are not private, only their constructors are.

File: lib/bson/regexp.rb [hunk 3]
```diff
@@ -196,15 +195,14 @@ def respond_to?(method, include_private = false)
...
       #
       # @return [ BSON::ByteBuffer ] The buffer with the encoded object.
       #
       # @see http://bsonspec.org/#/specification
       #
       # @since 4.2.0
-      def to_bson(buffer = ByteBuffer.new, validating_keys = Config.validating_keys?)
-        return compile.to_bson(buffer, validating_keys) if options.is_a?(Integer)
+      def to_bson(buffer = ByteBuffer.new)
+        return compile.to_bson(buffer) if options.is_a?(Integer)
```
Comment (p-mongo) [inline lib/bson/regexp.rb:205]: Didn't you remove integer options?

**Review by p-mongo — APPROVED** [review]

File: lib/bson/error/invalid_dbref_argument.rb (hunk 2 above)
Comment (Neilshweky) [reply lib/bson/error/invalid_dbref_argument.rb:7]: done

File: lib/bson/regexp.rb (hunk 3 above)
Comment (Neilshweky) [reply lib/bson/regexp.rb:205]: You're right, let me open a followup ticket.

File: lib/bson/error/invalid_dbref_argument.rb (hunk 2 above)
Comment (p-mongo) [reply lib/bson/error/invalid_dbref_argument.rb]: There are 2 or 3 other classes that this applies to.

Comment (Neilshweky) [reply lib/bson/error/invalid_dbref_argument.rb]: done

**Pull request merged**